src/Microsoft.Health.Fhir.Shared.Api/Modules/HealthChecks/HealthCheckConfiguration.cs

```diff
@@ -0,0 +1,49 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Linq;
+using EnsureThat;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Health.Fhir.Api.Modules.HealthChecks
+{
+    public class HealthCheckConfiguration : IPostConfigureOptions<HealthCheckServiceOptions>
+    {
+        private readonly IServiceProvider _serviceProvider;
+
+        public HealthCheckConfiguration(IServiceProvider serviceProvider)
+        {
+            EnsureArg.IsNotNull(serviceProvider, nameof(serviceProvider));
+
+            _serviceProvider = serviceProvider;
+        }
+
+        public void PostConfigure(string name, HealthCheckServiceOptions options)
```

Comment:
Is this method called before the ServiceProvider is materialized/build? I guess I don't know enough about the order of execution for this pipeline. Would there be a chance where the IHealthCheck instance is already created by the time it gets here?

Also, we could do this in the registration too right? Like at the very end of the registration, we can have a loop that loops through all ServiceDescriptor with service type of IHealthCheck and just overwrites that? Is there any advantage/disadvantage?

Reply:
After reviewing the framework code, this _seems_ like the best hook for what we're trying to do. This is called just after the container is built, but before the app starts, so it is safe in terms of ordering of registrations. I think this might be better than wrapping the registration in the container as it doesn't interfere with any of the HealthCheck extensions or HealthCheck lifetimes/scopes, these can all still be setup as desired.